sw/host/opentitanlib/src/crypto/spxplus.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+use std::io::{Read, Write};
+
+use anyhow::Result;
+
+use pqcrypto_sphincsplus::sphincsshake256128ssimple as spx;
+use pqcrypto_traits::sign::DetachedSignature;
+use pqcrypto_traits::sign::PublicKey;
+use pqcrypto_traits::sign::SecretKey;
+
+use crate::util::bigint::fixed_size_bigint;
+use crate::util::file::{FromReader, ToWriter};
+
+// Signature and key sizes taken from Table 8 on page 57 of the SPHINCS+ Round 3 Specification:
+// https://sphincs.org/data/sphincs+-round3-specification.pdf
+const PUBLIC_KEY_BYTE_LEN: usize = 32;
+const SECRET_KEY_BYTE_LEN: usize = 64;
+const SIGNATURE_BIT_LEN: usize = 7856 * 8;
+fixed_size_bigint!(Signature, at_most SIGNATURE_BIT_LEN);
+
+/// Trait for implementing public key operations.
+pub trait SpxPublicKeyPart {
+    /// Returns the public key component.
+    fn pk(&self) -> &spx::PublicKey;
+
+    fn pk_as_bytes(&self) -> &[u8] {
+        self.pk().as_bytes()
+    }
+
+    fn pk_len(&self) -> usize {
+        self.pk_as_bytes().len()
+    }
+
+    /// Verify a message signature, returning Ok(()) if the signature matches.
+    fn verify(&self, message: &[u8], sig: &SpxSignature) -> Result<()> {
+        spx::verify_detached_signature(
+            &spx::DetachedSignature::from_bytes(&sig.0.to_be_bytes())?,
+            message,
+            self.pk(),
+        )?;
+        Ok(())
+    }
+}
+
+/// A SPHINCS+ keypair consisting of the public and secret keys.
+#[derive(Clone)]
+pub struct SpxKeypair {
+    pk: spx::PublicKey,
+    sk: spx::SecretKey,
+}
+
+impl SpxKeypair {
+    /// Generates a new SPHINCS+ keypair.
+    pub fn generate() -> Self {
+        let (pk, sk) = spx::keypair();
+        SpxKeypair { pk, sk }
+    }
+
+    /// Sign `message` using the secret key.
+    pub fn sign(&self, message: &[u8]) -> SpxSignature {
+        let sm = spx::detached_sign(message, &self.sk);
+        SpxSignature(Signature::from_be_bytes(sm.as_bytes()).unwrap())
+    }
+
+    /// Consumes this keypair and returns the corrisponding public key.
+    pub fn into_public_key(self) -> SpxPublicKey {
+        SpxPublicKey(self.pk)
+    }
+}
+
+impl SpxPublicKeyPart for SpxKeypair {
+    fn pk(&self) -> &spx::PublicKey {
+        &self.pk
+    }
+}
+
+impl ToWriter for SpxKeypair {
+    fn to_writer(&self, w: &mut impl Write) -> Result<()> {
+        // Write out the keypair as a fixed length byte-string consisting of the public key
+        // concatenated with the secret key.
+        w.write_all(self.pk.as_bytes())?;
+        w.write_all(self.sk.as_bytes())?;
```

Comment:
Is the on-disk form of a Spx key or keypair just a bag of bytes? (ie: is there no file format?)

Should we consider a PEM container with a headers like "SPHINCS+ PUBLIC KEY" and the like?  The [pem_rfc7468 crate](https://crates.io/crates/pem-rfc7468) makes this sort of thing pretty easy:

```
// Turn your key into a bag of bytes:
let key = whatever_conversion_to_get_byte_buffer(); // (e.g. [u8; N] or Vec<u8>)
let pem = pem_rfc7468::encode_string("SPHINCS+ KEY PAIR", LineEnding::LF, &key)?;
w.write_all(pem.as_bytes())?;
```

Perhaps the decoding functionality is more useful, as you get back the "kind" so you can verify you've read the correct type of PEM file:

```
let mut data = Vec::new();
r.read_to_end(&mut data)?;
let (kind, key) = pem_rfc7468::decode_vec(&data)?;
if kind == "SPHINCS+ KEY PAIR" {
    Ok(SpxKeypair { ... })
} else {
    Err("The wrong kind of thing");
}
```

Reply:
I agree that a PEM file is much nicer than the bag of bytes. It's my understanding that there is no format for SPHINCS+ keys, and that the format isn't all that important at the moment, so long as opentitantool is consistent. I'm in favor of this, it seems like a much better approach.

Reply:
There is no standard format yet. We floated binary blobs, hjson files (which can also encode type and other info) but PEM SGTM too. I'd be OK with addressing this in a follow-up PR.